Feat: PIM factories (restricted and immutable)

fabianschu opened the pull request:
This PR consolidates the efforts on the two PIM factories:
* it removes the outdated generalized factory
* it adds the restricted factory (Qacc use case)
* it adds the immutable factory (coin launcher use case)

fabianschu committed:
chore: removes generalized factory

fabianschu committed:
chore: adds restricted factory

fabianschu committed:
chore: adds interface for curve interaction role

fabianschu committed:
chore: adds restricted curve setup to e2e reg

fabianschu committed:
chore: adds immutable factory

fabianschu committed:
chore: adds NatSpec for immutable factory

fabianschu committed:
chore: adds more comments

fabianschu committed:
chore: adds deploy scripts for both factories

fabianschu committed:
chore: use generalized deployment script

fabianschu committed:
chore: formatting

fabianschu commented:
Open questions for me at this point:

Restricted PIM Factory:
* right now all privileges that exist go to the address that is set as `initialAdmin`
* this comprises:
  * being workflow admin (1)
  * receiving initialIssuanceSupply (2)
  * being token owner for issuance token (3)
  * receiving CURVE_INTERACTION_ROLE (4) 

However, I learned this might not be ideal for qacc, since the participating projects won't be orchestrator admins. Would it make sense to make a "split" between (1) and (2,3,4)? So the user passes another address to the factory that receives privileges (2,3,4) while the address passed as `initialAdmin` "only" receives privilege (1)? @0xNuggan

marvinkruse commented on script/deploymentScript/DeployWorkflowSpecificFactory.s.sol:
```diff
@@ -0,0 +1,55 @@
+                console2.log(
+                    "Deploying Restricted_PIM_Factory_v1 at address: ",
+                    address(
+                        new Restricted_PIM_Factory_v1(
+                            orchestratorFactory, trustedForwarder
```
I could be wrong, but this would also execute if both addresses (or at least one) is `address(0)`, right? Maybe we should add something like we have in the `DeploymentScript.s.sol` to ensure that the deployment is gonna be okay? (see [here](https://github.com/InverterNetwork/contracts/blob/2a8a4c80ff4f24a59546d4e6126b81bc51228c94/script/deploymentScript/DeploymentScript.s.sol#L458))

fabianschu replied:
Great idea!

marvinkruse commented on src/factories/interfaces/IImmutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,91 @@
+    //--------------------------------------------------------------------------
+    // Functions
+
+    /// @notice Deploys a new issuance token and uses that to deploy a workflow with an open yet immutable bonding curve.
```
I believe @Zitzak reformatted our natspec to also adhere to a specific max length, did you double check this here already? otherwise we might wanna do that

fabianschu replied:
Yeah, looks a bit long, I shortened that instance but I'm not sure where I can find the actual max length. @Zitzak can you help out

Zitzak replied:
I believe I followed this guideline
https://docs.soliditylang.org/en/latest/style-guide.html#maximum-line-length

marvinkruse commented on src/factories/interfaces/IImmutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,91 @@
+    /// @param moduleConfigs Variable length set of optional module's config data.
+    /// @param issuanceTokenParams The issuance token's parameters (name, symbol, decimals, maxSupply).
+    /// @param initialPurchaseAmount The volume of the first purchase in terms of collateral token.
+    /// @return CreatedOrchestrator Returns the created orchestrator instance.
```
I believe the first letter is usually lower-case? `createdOrchestrator`

marvinkruse replied:
Also not changed yet, just fyi.

marvinkruse commented on src/factories/workflow-specific/Immutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,184 @@
+// External Dependencies
+import {ERC2771Context, Context} from "@oz/metatx/ERC2771Context.sol";
+
+contract Immutable_PIM_Factory_v1 is
```
Missing the header natspec (title, description, security notice)

marvinkruse commented on src/factories/workflow-specific/Immutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,184 @@
+    //--------------------------------------------------------------------------
+    // State Variables
+
+    /// @dev	store address of {Orchestratorfactory_v1}.
```
```suggestion
    /// @dev	Store address of {Orchestratorfactory_v1}.
```

marvinkruse replied:
Hasn't changed yet, if I'm not mistaking

marvinkruse commented on src/factories/workflow-specific/Immutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,184 @@
+    /// @dev	store address of {Orchestratorfactory_v1}.
+    address public orchestratorFactory;
+
+    /// @dev	mapping of Funding Manager address to `feeRecipient` address.
```
```suggestion
    /// @dev	Mapping of Funding Manager address to `feeRecipient` address.
```

marvinkruse commented on src/factories/workflow-specific/Immutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,184 @@
+            issuanceTokenParams.maxSupply,
+            address(this) // assigns owner role to itself initially to manage minting rights temporarily
+        );
+        // MODIFY AUTHORIZER CONFIG
```
A line break between logical sections would make sense IMHO, so here before this like (`// MODIFY AUTHORIZER CONFIG`), then again before `// MODIFY AUTHORIZER CONFIG`, etc.

marvinkruse reviewed (changes requested):
Good stuff in general, nice! 🙌 

(I haven't fully looked through the tests yet, just glossed over them, but the rest I reviewed.)

fabianschu committed:
chore: sends issuance supply to deployer

fabianschu committed:
feat: adds `addFunding` method

fabianschu committed:
feat: adds `withdrawFunding`

fabianschu committed:
chore: adds more tests

fabianschu committed:
chore: adds comments & NatSpec

fabianschu committed:
chore: fmt

fabianschu committed:
chore: validates inputs in deployment script

fabianschu committed:
chore: marvin feedback I

fabianschu committed:
chore: fmt

fabianschu commented:
I addressed your comments @marvinkruse. 
In addition to that I had to make some additional changes to cater for qacc specific needs:
* the restricted factory now contains a pre-funding mechanism where anyone can sponsor the initial collateral supply required for the deployment of a PIM workflow (sponsor can also withdraw their pre-funding again)
* to be able to use pre-funding, deployer needs to set sponsor as workflow admin
* initial issuance supply goes to deployer
* curve interaction role also goes to deployer

Note: fee claim for deployer is planned to be enabled through admin's Multisig via Zodiac Roles

fabianschu committed:
chore: feedback Marvin II

0xNuggan commented on src/factories/interfaces/IImmutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,109 @@
+    /// @param issuanceToken The address of the issuance token.
+    /// @param deployer The address of the deployer.
+    event PIMWorkflowCreated(
+        address indexed fundingManager,
```
I guess this is for indexing reasons, but why reuturn the fundingManager instead of the Orchestrator?

0xNuggan replied:
The implementation returns the orchestrator address, so thi is a bug :)

fabianschu replied:
woopsies

0xNuggan commented on src/factories/interfaces/IRestricted_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,108 @@
+    /// @param token Address of token used for funding.
+    /// @param amount Funding amount.
+    event FundingAdded(
+        address indexed paymaster,
```
Not sure I like calling it "paymaster" since that term is used in account abstraction. Maybe "sponsor" is better?

fabianschu committed:
fix: event param

fabianschu committed:
chore: rename `paymaster` => `sponsor`

marvinkruse commented on src/factories/interfaces/IImmutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,109 @@
+import {ERC20Issuance_v1} from "src/external/token/ERC20Issuance_v1.sol";
+
+/**
+ * @title   Immutable PIM Factory
```
Our usual naming includes the word "Inverter" in the `@title` now, so "Inverter Payment Router" instead of just "Payment Router". I think we should be consistent here as well then?

"Immutable Inverter PIM Factory" or so? And for Restricted as well.

marvinkruse commented on src/factories/workflow-specific/Immutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,184 @@
+        external
+        onlyPimFeeRecipient(fundingManager)
+    {
+        _pimFeeRecipients[fundingManager] = to;
```
Saw the thumbs-up here but nothing has changed, right? Which is fine, just wanna make sure that we don't forget about it, in case you intended to add something here. If you think this should be left like it is, that's also fine.

fabianschu replied:
Sorry, I overlooked this. Added a check now.

marvinkruse commented on src/factories/workflow-specific/Restricted_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,186 @@
+// External Dependencies
+import {ERC2771Context, Context} from "@oz/metatx/ERC2771Context.sol";
+
+contract Restricted_PIM_Factory_v1 is
```
You added the Natspec headers to the interfaces, instead of to the contracts. We have them on the contracts, not the interfaces so far. 
Yesterday in the dev guidelines call we decided we want to have it in both actually (as I also see that having it in the interface makes a lot of sense). We will then (later) add the functionality to the CI bot to verify that both interface and contract have the same Natspec header (just so we don't forget to add it for one of them or if we change it but don't update the other one). Does that make sense?

marvinkruse commented on src/factories/workflow-specific/Restricted_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,238 @@
+        // enable bonding curve to mint issuance token
+        issuanceToken.setMinter(fundingManager, true);
+
+        // transfer initial collateral supply from msg.sender to  bonding curve and mint issuance token to recipient
```
Double space between `to` and `bonding curve`.

```suggestion
        // Transfer initial collateral supply from msg.sender to bonding curve and mint issuance token to recipient
```

0xNuggan replied:
Just out of curiosity, @marvinkruse : Do you have some kind of plugin that highlights these kind of style mistakes? I would never have noticed that double space :rofl:

fabianschu replied:
I just removed the whitespace but didn't capitalize the first letter for consistency :P

marvinkruse commented on src/factories/workflow-specific/Restricted_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,238 @@
+    mapping(
+        address sponsor
+            => mapping(address actor => mapping(address token => uint amount))
+    ) fundings;
```
Missing the visibility keyword.

fabianschu replied:
I made this public.

marvinkruse commented on src/factories/workflow-specific/Restricted_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,238 @@
+        fundings[admin][actor][address(collateralToken)] -=
+            initialCollateralSupply;
+
+        // collateral token is paid for by the msg.sender
```
This comment is confusing me slightly, as we are using the tokens that are deposited by the sponsor, not necessarily the msg.sender?

0xNuggan replied:
yes, it should say either sponsor or factory

fabianschu replied:
yea, outdated comment is fixed

marvinkruse commented on src/factories/workflow-specific/Restricted_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,238 @@
+        address actor
+    ) private {
+        uint availableFunding = fundings[admin][actor][address(collateralToken)];
+        if (availableFunding < initialCollateralSupply) {
```
As it reverts here this basically means that it's not an option to pay this as the deployer in that specific moment, there needs to be a sponsor that pre-funded, right? Yes, the actor can be the sponsor, but this entails two calls (first `addFunding` before creating the workflow. Is that intentional? In my mind a fallback mechanism would make sense where the sponsors funds are taken, and if not enough, the deployer is paying (if the approved it)? But if that's not required for the client or I am just missing something then feel free to ignore this ofc.

0xNuggan replied:
In this case the client will pre-seed the contract with the funds, since projects aren't supposed to pay anything. A fallback mechanism makes sense, but is somewhat out of scope for this case?

fabianschu replied:
If you don't mind I'll keep it qacc-specific and refrain from adding the fallback.

marvinkruse commented on script/deploymentScript/DeployWorkflowSpecificFactory.s.sol:
```diff
@@ -0,0 +1,55 @@
+pragma solidity ^0.8.0;
+
+import "forge-std/Script.sol";
+import "forge-std/console.sol";
```
Got a 👍 but no changes, was that intentional? (just double checking)

fabianschu replied:
No, not intentional. Done now :)

fabianschu committed:
chore: adds PR feedback

fabianschu committed:
chore: fmt

marvinkruse commented:
@fabianschu Overall looks good to me, one thing I just noticed though: we want to deploy the factories any time a testnet deployment goes through, potentially even with every deployment (but I'm not sure about that last part). Two things come to mind there: (a) it should be included in the TestnetDeploymentScript to automatically just deploy both factories and (b) for that it would probably be cool to have not just a `run()` function in the `DeployWorkflowSpecificFactory.s.sol` but also a run(orchestratorFactory, trustedForwarder, factoryType)` or something like that - if that makes sense?

fabianschu committed:
chore: deploy specific factories in general testnet deployment script

fabianschu committed:
chore: rename workflow-specific => custom

fabianschu committed:
chore: make fmt

fabianschu committed:
Feat: Mint wrapper (#663)

* feat: adds `MintWrapper` contract

* chore: removes trusted forwarder from mint wrapper

* feat: uses mint wrapper in factory

* chore: adds more documentation

---------

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

marvinkruse committed:
Merge branch 'dev' into feat/pim-factories-restricted-and-immutable

marvinkruse committed:
Script: Streamline deployment script

0xNuggan commented on src/factories/custom/Immutable_PIM_Factory_v1.sol:
```diff
@@ -0,0 +1,210 @@
+
+    /// @dev	Modifier to guarantee the caller is the fee recipient for the given funding manager.
+    modifier onlyPimFeeRecipient(address fundingManager) {
+        if (_msgSender() != _pimFeeRecipients[fundingManager]) {
+            revert PIM_WorkflowFactory__OnlyPimFeeRecipient();
+        }
```
Not gonna be an issue here, but for consistency

0xNuggan reviewed (changes requested):
just two small things

0xNuggan reviewed (changes requested):
Just two small things

fabianschu committed:
chore: separates deployer from beneficiary (#671)

* chore: separates deployer from beneficiary

* fix: comments

fabianschu committed:
chore: adds symbol & name to MintWrapper

fabianschu commented:
Will be split into multiple PRs
